File: langs/java/src/main/java/dev/dotlyte/BoundaryConfig.java
package dev.dotlyte;

import java.util.*;
import java.util.function.BiConsumer;

/**
 * Boundary-aware configuration that partitions keys into <em>server</em>,
 * <em>client</em>, and <em>shared</em> sets.
 *
 * <p>A {@code BoundaryConfig} wraps a plain data map and provides filtered
 * views. Because Java always runs on the server, {@link #isServerContext()}
 * returns {@code true} and {@link #isClientContext()} returns {@code false}.
 *
 * <pre>{@code
 * Set<String> serverKeys = Set.of("DB_PASSWORD", "API_SECRET");
 * Set<String> clientKeys = Set.of("APP_NAME", "THEME");
 * Set<String> sharedKeys = Set.of("LOG_LEVEL");
 *
 * BoundaryConfig bc = new BoundaryConfig(data, serverKeys, clientKeys, sharedKeys);
 * Map<String, Object> client = bc.clientOnly();
 * }</pre>
 */
public final class BoundaryConfig {

    private final Map<String, Object> data;
    private final Set<String> serverKeys;
    private final Set<String> clientKeys;
    private final Set<String> sharedKeys;
    private BiConsumer<String, String> onSecretAccess;

    /**
     * Create a boundary-aware configuration.
     *
     * @param data       the underlying configuration values
     * @param serverKeys keys that are only available on the server
     * @param clientKeys keys that are only available on the client
     * @param sharedKeys keys available in both contexts
     * @throws NullPointerException if any argument is {@code null}
     */
    public BoundaryConfig(
            final Map<String, Object> data,
            final Set<String> serverKeys,
            final Set<String> clientKeys,
            final Set<String> sharedKeys) {

        Objects.requireNonNull(data, "data must not be null");
        Objects.requireNonNull(serverKeys, "serverKeys must not be null");
        Objects.requireNonNull(clientKeys, "clientKeys must not be null");
        Objects.requireNonNull(sharedKeys, "sharedKeys must not be null");

        this.data = Collections.unmodifiableMap(new LinkedHashMap<>(data));
        this.serverKeys = Collections.unmodifiableSet(new LinkedHashSet<>(serverKeys));
        this.clientKeys = Collections.unmodifiableSet(new LinkedHashSet<>(clientKeys));
        this.sharedKeys = Collections.unmodifiableSet(new LinkedHashSet<>(sharedKeys));
    }

    /**
     * Set a callback that is invoked when a server-only key is accessed.
     *
     * @param onSecretAccess callback receiving {@code (key, value)}
     */
    public void setOnSecretAccess(final BiConsumer<String, String> onSecretAccess) {
        this.onSecretAccess = onSecretAccess;
    }

    // ── Access ─────────────────────────────────────────────────

    /**
     * Retrieve a configuration value, respecting boundary rules.
     *
     * <p>If the key belongs to {@code serverKeys} and we are in a server
     * context, the secret-access callback is invoked (if set).
     *
     * @param key the configuration key
     * @return the value, or {@code null} if the key is not present
     * @throws DotlyteException if the key is not part of any boundary set
     */
    public Object get(final String key) {
        if (!serverKeys.contains(key) && !clientKeys.contains(key)
                && !sharedKeys.contains(key)) {
            throw new DotlyteException(
                    "Key '" + key + "' is not in any boundary set (server, client, shared).", key);
        }

        final Object value = data.get(key);

        if (serverKeys.contains(key) && value != null && onSecretAccess != null) {
            onSecretAccess.accept(key, String.valueOf(value));
        }

        return value;
    }

    // ── Filtered views ─────────────────────────────────────────

    /**
     * Return an unmodifiable map containing only server and shared keys.
     *
     * @return server-only view
     */
    public Map<String, Object> serverOnly() {
        return filterByKeys(serverKeys, sharedKeys);
    }

    /**
     * Return an unmodifiable map containing only client and shared keys.
     *
     * @return client-only view
     */
    public Map<String, Object> clientOnly() {
        return filterByKeys(clientKeys, sharedKeys);
    }

    // ── Context queries ────────────────────────────────────────

    /**
     * Always returns {@code true} because Java runs on the server.
     *
     * @return {@code true}
     */
    public static boolean isServerContext() {
        return true;
    }

    /**
     * Always returns {@code false} because Java runs on the server.
     *
     * @return {@code false}
     */
    public static boolean isClientContext() {
        return false;
    }

    // ── Immutability enforcement ───────────────────────────────

    /**
     * Not supported — BoundaryConfig is immutable.
     *
     * @throws UnsupportedOperationException always
     */
    public void put(final String key, final Object value) {
        throw new UnsupportedOperationException("BoundaryConfig is immutable");
    }

    /**
     * Not supported — BoundaryConfig is immutable.
     *
     * @throws UnsupportedOperationException always
     */
    public void remove(final String key) {
        throw new UnsupportedOperationException("BoundaryConfig is immutable");
    }

    // ── Helpers ────────────────────────────────────────────────

    private Map<String, Object> filterByKeys(final Set<String> primary, final Set<String> shared) {
        final Map<String, Object> result = new LinkedHashMap<>();
        for (final Map.Entry<String, Object> entry : data.entrySet()) {
            if (primary.contains(entry.getKey()) || shared.contains(entry.getKey())) {
                result.put(entry.getKey(), entry.getValue());
            }
        }
        return Collections.unmodifiableMap(result);
    }
}

File: langs/php/src/BoundaryConfig.php
<?php

declare(strict_types=1);

namespace Dotlyte;

/**
 * Boundary-aware configuration for server/client key separation.
 *
 * Enforces access controls so that server-only keys are never
 * accidentally leaked to client contexts.
 */
final class BoundaryConfig
{
    /** @var array<string, mixed> */
    private array $data;

    /** @var list<string> */
    private array $serverKeys;

    /** @var list<string> */
    private array $clientKeys;

    /** @var list<string> */
    private array $sharedKeys;

    /** @var callable|null */
    private $onSecretAccess;

    /**
     * @param array<string, mixed> $data        All config data
     * @param list<string>         $serverKeys  Keys restricted to server context
     * @param list<string>         $clientKeys  Keys allowed in client context
     * @param list<string>         $sharedKeys  Keys available in both contexts
     * @param callable|null        $onSecretAccess Called with (key, value) on access
     */
    public function __construct(
        array $data,
        array $serverKeys,
        array $clientKeys,
        array $sharedKeys = [],
        ?callable $onSecretAccess = null,
    ) {
        $this->data = $data;
        $this->serverKeys = array_values($serverKeys);
        $this->clientKeys = array_values($clientKeys);
        $this->sharedKeys = array_values($sharedKeys);
        $this->onSecretAccess = $onSecretAccess;
    }

    /**
     * Get a config value with boundary checking.
     *
     * In PHP (always server context) all keys are accessible, but a
     * callback is triggered for sensitive access tracking.
     *
     * @throws DotlyteException If the key is not declared in any boundary list
     */
    public function get(string $key): mixed
    {
        $allDeclared = array_merge($this->serverKeys, $this->clientKeys, $this->sharedKeys);

        if (!empty($allDeclared) && !in_array($key, $allDeclared, true)) {
            throw new DotlyteException(
                "Key '{$key}' is not declared in any boundary (server, client, shared). "
                . 'Add it to the appropriate boundary list.',
                key: $key,
            );
        }

        $value = $this->resolveKey($key);

        if ($this->onSecretAccess !== null && in_array($key, $this->serverKeys, true)) {
            ($this->onSecretAccess)($key, $value);
        }

        return $value;
    }

    /**
     * Return only keys designated as server-only, filtered from data.
     *
     * @return array<string, mixed>
     */
    public function serverOnly(): array
    {
        return $this->filterKeys($this->serverKeys);
    }

    /**
     * Return only keys designated as client-safe, filtered from data.
     *
     * @return array<string, mixed>
     */
    public function clientOnly(): array
    {
        return $this->filterKeys($this->clientKeys);
    }

    /**
     * PHP always runs on the server.
     */
    public function isServerContext(): bool
    {
        return true;
    }

    /**
     * Resolve a dot-notation key from the data array.
     */
    private function resolveKey(string $key): mixed
    {
        if (array_key_exists($key, $this->data)) {
            return $this->data[$key];
        }

        // Try dot-notation traversal
        $parts = explode('.', $key);
        $current = $this->data;

        foreach ($parts as $part) {
            if (!is_array($current) || !array_key_exists($part, $current)) {
                return null;
            }
            $current = $current[$part];
        }

        return $current;
    }

    /**
     * Filter the data array to only include the specified keys.
     *
     * @param list<string> $keys
     * @return array<string, mixed>
     */
    private function filterKeys(array $keys): array
    {
        $result = [];
        foreach ($keys as $key) {
            $value = $this->resolveKey($key);
            if ($value !== null) {
                $result[$key] = $value;
            }
        }
        return $result;
    }
}
